src/testpool.jl

```diff
@@ -146,40 +137,120 @@ resize_test_engine_pool!(10, id->"RAITest-test-\$id")
 resize_test_engine_pool!(0)
 ```
 """
-function resize_test_engine_pool!(size::Int64, generator::Option{Function}=nothing)
+function resize_test_engine_pool!(size::Int64, name_generator::Option{Function}=nothing)
     if size < 0
         size = 0
     end
 
-    if !isnothing(generator)
-        TEST_ENGINE_POOL.generator = generator
+    if !isnothing(name_generator)
+        TEST_ENGINE_POOL.name_generator = name_generator
     end
 
+    @lock TEST_SERVER_LOCK begin
```

Comment:
If I'm understanding the concurrency model correctly (which is a big if to be fair), the main place where we could see data update races is when marking engines as available for testing or not. It looks like all of the pool maintenance isn't really threaded. If this is indeed the case, could we use the lock just on those availability updates or use an atomic there?

Reply:
Pool maintenance can be performed while testing is in progress. This is mainly useful for replacing engines while tests are being run (which we've never actually done as they've never failed during testing). The single global lock for adding/removing/claiming engines is the simplest, most reliable approach to allow this.

If we're not modifying the list during testing (and we aren't in current use) then we could add atomics, and drop the add/remove lock entirely, marking those methods as non-threadsafe.

So, yes we could change it. Is there an advantage in reliability or readability either way?

Reply:
Not a huge one, but it seems like there are some opportunities for helpers that could be simplified without the lock or would require smaller chunks of code locking. For instance, we delete engines in a similar try/catch + warn pattern in a bunch of places that could maybe be abstracted a bit

Reply:
Modifying engines mid-testing didn't prove to be needed, so simplifying the threading model sounds reasonable to me. I'll do this in a followup PR so it's easier to see the changes.

Reply:
I found potential issues so I went ahead and made the changes in this PR. More work is now done by helpers, and the locks are held for less of the code.